File: week1/11.container_with_most_water/11.container_with_most_water.go
// O(n)
func maxArea(a []int) int {
	area, max, left, right := 0, 0, 0, len(a)-1
	for left < right {
		if a[left] < a[right] {
			area = (right - left) * a[left]
			left++
		} else {
			area = (right - left) * a[right]
			right--
		}

		if max < area {
			max = area
		}
	}
	return max
}

File: week1/11.container_with_most_water/11.container_with_most_water.js
// O(n^2)
var maxArea = function (a) {
    let max = 0;

    for (let i = 0; i < a.length - 1; i++) {
        for (let j = i + 1; j < a.length; j++) {
            let area = (j - i) * Math.min(a[i], a[j]);
            max = area > max ? area : max;
        }
    }
    return max;
};


// O(n)

var maxArea = function (a) {
    let max = 0;
    for (let i = 0, j = a.length - 1; i < j;) {
        let minH = a[i] < a[j] ? a[i++] : a[j--];
        max = Math.max(max, minH * (j - i + 1));
    }
    return max;
}